Introduce the transient files engine

Konamiman opened the pull request:
### Changes proposed in this Pull Request:

This pull request introduces a transient files engine intended for registering files that have an expiration date and can be served via a public, unauthenticated HTTP URL. Initially, the only mechanism provided to create these files is providing the raw contents of the file via a input string; but additional mechanisms could be added in the future, such as directly providing the path of an existing file.

See [the README file](https://github.com/woocommerce/woocommerce/blob/add-transient-files-engine/plugins/woocommerce/src/Internal/TransientFiles/README.md) and the documentation comments in [the TransientFilesEngine class](https://github.com/woocommerce/woocommerce/blob/add-transient-files-engine//plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php) for all the nitty-gritty details.

This pull request is a simplified version of https://github.com/woocommerce/woocommerce/pull/41055.


### How to test the changes in this Pull Request:

### Testing the code API

1. Run the following:

```
wp eval "echo wc_get_container()->get(\Automattic\WooCommerce\Internal\TransientFiles\TransientFilesEngine::class)->create_transient_file('foobar', '2024-01-01');"
```

You'll get a file name whose first 6 digits is the expiration date encoded in hexadecimal (so `7e8 1 01`).

2. Run the following to see the exact physical path of the generated file, where `<file name>` is the result from the previous call:

```
wp eval "echo wc_get_container()->get(\Automattic\WooCommerce\Internal\TransientFiles\TransientFilesEngine::class)->get_transient_file_path('<file name>');"
```

You should have got `<WordPress directory>/wp-content/uploads/woocommerce_transient_files/2024-01-01/<file name minus expiration date prefix>`. Verify that the file actually exists and it contains the string `foobar`.

5. Run the following to delete the generated file:

```
wp eval "var_export(wc_get_container()->get(\Automattic\WooCommerce\Internal\TransientFiles\TransientFilesEngine::class)->delete_transient_file('<file name>'));"
```

You'll get `true` and the file (and also the `2024-01-01` directory) will have disappeared. If you repeat the call you'll get `false`.

### Testing the expired files deletion

To easily test the expired files deletion mechanism we need to be able to create transient files that have already expired. This is not possible out of the box, but can be achieved easily by manually altering the created directory name. So go ahead, create 10 files by repeating step 1 as many times and then rename the  `2024-01-01` directory to `2023-01-01` (and now if you wanted to actually locate these files by using `get_transient_file_path` you'd have to change the first digits of the expiration date prefix from `7e8` to `7e7` in the passed file name).

Now run the following:

```php
wp eval "var_export(wc_get_container()->get(\Automattic\WooCommerce\Internal\TransientFiles\TransientFilesEngine::class)->delete_expired_files(7));"
```

The result will be:

```php
array (
  'deleted_count' => 7,
  'files_remain' => true,
)
```

and there will be only 3 files remaining in the `2023-01-01` directory.

Running the same command again will yield the following:

```php
array (
  'deleted_count' => 3,
  'files_remain' => false,
)
```

and the `2023-01-01` directory will have disappeared.

You can also try the scheduled action by going to WooCommerce - Status - Tools and re-scheduling the cleanup:

![image](https://github.com/woocommerce/woocommerce/assets/937723/46694948-7a79-4f77-8e4b-f579766b945a)

By rescheduling the cleanup the action will be scheduled to keep running immediately as long as there's still data to delete, and then every 24h. You can change that interval with a filter, e.g. every 10 seconds:

```php
add_filter('woocommerce_delete_expired_transient_files_interval', fn($interval) => 10);
```

Note however that the scheduled action deletes 1000 files every time, so for testing you'll need to either create as many transient files or tweak the code to use a smaller batch (pass the batch size to the `delete_expired_files` call in `TransientFilesEngine::handle_expired_files_cleanup_action`).


### Testing the HTTP endpoint

Generate a transient file by repeating step 1 and verify that you can access the rendered file in your browser (even from an incognito window) or using any tool able to perform HTTP requests (replace `<filename>` with the file name that you get from `create_transient_file`): `http://localhost/wc/file/transient/<filename>`.

If you rename the folder to `2023-01-01` and try to access the file replacing `7e8` with `7e7` at the beginning of its name, you'll get a 404 error, since the endpoint doesn't serve files that have expired (but you can still get the file path with `get_transient_file_path`).


### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

Konamiman committed:
Add the TransientFilesEngine class.

Also add a migration to create the base directory for transient files
at WooCommerce install time.

Konamiman committed:
Add unit tests for the TransientFilesEngine class.

Konamiman committed:
Add documentation for the transient files engine.

Konamiman committed:
Add changelog file

Konamiman committed:
Small fixes

coreymckrill commented on plugins/woocommerce/includes/class-woocommerce.php:
```diff
@@ -263,6 +263,7 @@ private function init_hooks() {
 		$container->get( FeaturesController::class );
 		$container->get( WebhookUtil::class );
 		$container->get( Marketplace::class );
+		$container->get(\Automattic\WooCommerce\Utilities\TimeUtil::class);
```
Maybe worth adding a `use` statement at the top so you don't have to include the full namespace here? It would be more consistent with the rest of the classes in this list.

coreymckrill commented on plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php:
```diff
@@ -0,0 +1,496 @@
+
+		$realpathed_transient_files_directory = $this->legacy_proxy->call_function( 'realpath', $transient_files_directory );
+		if ( false === $realpathed_transient_files_directory ) {
+			throw new Exception( "The base transient files directory doesn't exist: $transient_files_directory" );
```
Instead of creating the default directory in `wc_update_850_create_transient_files_directory` during the upgrade routine, could you just create it here if it doesn't exist? This way you could potentially avoid creating a directory that would never be used if the user adds a filter to customize the directory name.

Konamiman replied:
A different directory to hold the transient files can be specified via filter, and for security and robustness I'm leaving the responsibility to create that directory to the filterer and throwing an error if it doesn't exist. So for consistency I'm doing the same with the default directory, and this implies that it needs to be created "externally" at some point.

coreymckrill commented on plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php:
```diff
@@ -0,0 +1,496 @@
+	public function get_transient_file_path( string $filename ): ?string {
+		if ( strlen( $filename ) < 7 ) {
+			return null;
+		}
```
When I was going through the testing instructions, I forgot to replace `<file name>` with the actual file name the first time I ran the command that calls this method. I got a couple of deprecation notices:

```
Deprecated: Invalid characters passed for attempted conversion, these have been ignored in /var/www/html/wp-content/plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php on line 172
Deprecated: Invalid characters passed for attempted conversion, these have been ignored in /var/www/html/wp-content/plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php on line 173
```

Just thinking it might be worth adding some more validation here to make sure the filename string only contains hex-compatible characters.

coreymckrill commented on plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php:
```diff
@@ -0,0 +1,496 @@
+
+			if ( $this->file_has_expired( $file_path ) ) {
+				$legacy_proxy->call_function( 'status_header', 404 );
+				$legacy_proxy->exit();
```
When I try getting a file that either doesn't exist or is expired, the browser screen is just blank (I can see the 404 status if I open the dev console). 🤔 I wonder if it would be helpful here to use a `wp_die()` with a message like "File expired." Or, to be ambiguous about whether the file ever existed, "File not found." What do you think?

Konamiman replied:
These files are (in principle) not intended for human consumption: for the receipts use case at least, the URL of the file will be handled to an automated printing service. So I think it's clearer to get no content at all than to get a (seemingly arbitrary) error message (taken to the extreme, a badly implemented printing service not printing anything at all is better than it printing the string "File not found").

coreymckrill replied:
> a badly implemented printing service not printing anything at all is better than it printing the string "File not found"

😄 It is all too easy to envision this exact thing happening, so you have a point well taken.

coreymckrill commented on plugins/woocommerce/src/Internal/TransientFiles/README.md:
```diff
@@ -0,0 +1,117 @@
+
+## Summary
+
+The WooCommerce transient files engine allows to create transient files. We define a _transient file_ as a physical file (in the server filesystem) that:
```
```suggestion
The WooCommerce transient files engine allows for the creation of transient files. We define a _transient file_ as a physical file (in the server filesystem) that:
```

coreymckrill commented on plugins/woocommerce/src/Internal/TransientFiles/README.md:
```diff
@@ -0,0 +1,117 @@
+* Can have any arbitrary content.
+* Lives in a dedicated subdirectory inside the WordPress `uploads` folder (customizable via [hook](#hooks)).
+* Has an expiration date.
+* Its contents can be retrieved via [unauthenticated HTTP endpoint](#the-http-file-serving-endpoint).
```
```suggestion
* Its contents can be retrieved via an [unauthenticated HTTP endpoint](#the-http-file-serving-endpoint).
```

coreymckrill commented on plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php:
```diff
@@ -0,0 +1,496 @@
+		$filename = bin2hex( $this->legacy_proxy->call_function( 'random_bytes', 16 ) );
+
+		$transient_files_directory  = $this->get_transient_files_directory();
+		$transient_files_directory .= '/' . substr( $expiration_date, 0, 10 );
```
```suggestion
		$transient_files_directory .= '/' . $expiration_date_object->format( 'Y-m-d' );
```

coreymckrill commented on plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php:
```diff
@@ -0,0 +1,496 @@
+		$filepath = $transient_files_directory . '/' . $filename;
+
+		// phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_read_file_put_contents
+		if( false === $this->legacy_proxy->call_function( 'file_put_contents', $filepath, $file_contents ) ) {
```
Why not use the `$wp_filesystem` method here, and avoid adding a linting exception?

coreymckrill commented on plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php:
```diff
@@ -0,0 +1,496 @@
+	 * 3. Otherwise, use 'file_has_expired' passing the obtained full file path.
+	 *
+	 * @param string $file_path The full file path to check.
+	 * @return bool Trye if the file has expired, false otherwise.
```
```suggestion
	 * @return bool True if the file has expired, false otherwise.
```

coreymckrill commented on plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php:
```diff
@@ -0,0 +1,496 @@
+	 */
+	public function file_has_expired( string $file_path ): bool {
+		$dirname                = dirname( $file_path );
+		$expiration_date        = substr( $dirname, strlen( $dirname ) - 10, 10 );
```
```suggestion
		$expiration_date        = basename( $dirname );
```

coreymckrill commented on plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php:
```diff
@@ -0,0 +1,496 @@
+			$expiration_date = gmdate( 'Y-m-d', $expiration_date );
+		} elseif ( ! is_string( $expiration_date ) || ! TimeUtil::is_valid_date( $expiration_date, 'Y-m-d' ) ) {
+			$expiration_date = is_scalar( $expiration_date ) ? $expiration_date : gettype( $expiration_date );
+			throw new InvalidArgumentException( "$expiration_date is not a valid date, expected format: year-month-day" );
```
```suggestion
			throw new InvalidArgumentException( "$expiration_date is not a valid date, expected format: YYYY-MM-DD" );
```

This seems a bit less ambiguous about the expected format.

coreymckrill reviewed (commented):
👍 Overall this looks really good! Clear testing instructions, and everything worked as described. I left a few wording suggestions and a couple of minor questions to discuss.

Konamiman committed:
Small fixes:

- Introduce using statement
- Use WP_Filesystem instead of file_put_contents
- Validate that filename contains only hex digits in get_transient_file_path
- Replace year-month-day with YYYY-MM-DD in error message

Konamiman committed:
Apply suggestions from code review

Co-authored-by: Corey McKrill <916023+coreymckrill@users.noreply.github.com>

Konamiman committed:
Merge branch 'trunk' into add-transient-files-engine

Konamiman commented:
@coreymckrill There's a new commit that manually applies some of your suggestions, and another one that applied others directly from GitHub.

coreymckrill reviewed (approved):
Thanks, this looks great! I'm going to try re-running the tests to see if we can get them to pass this time, the failures don't look related to this at all. Might also need to do a rebase, though.

Konamiman committed:
Merge branch 'trunk' into add-transient-files-engine

Konamiman committed:
Create the default base directory for transients as needed, not in a migration.

If the woocommerce_transient_files_directory is used and the
directory it returns doesn't exist, an exception will be thrown.
If the filter is not used and the default base directory doesn't exist,
it will be created by get_transient_files_directory.

Konamiman committed:
Fix linting errors

jazir555 commented:
Will the use of eval will cause security issues here?

Konamiman commented:
> Will the use of eval cause security issues?

Sorry @jazir555, what are you referring to? `eval` is not used anywhere in the code introduced in this pull request.

Konamiman commented:
> This is in the opening comment of the post, is this just to test the feature functionality?

Yes, it's just a command line way to test the feature, the code itself is not using the `eval` function.
